services/QuillLMS/app/models/teacher_notification.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: teacher_notifications
+#
+#  id                :bigint           not null, primary key
+#  email_sent        :datetime
+#  message_attrs     :jsonb
+#  notification_type :text             not null
+#  created_at        :datetime         not null
+#  updated_at        :datetime         not null
+#  user_id           :bigint           not null
+#
+# Indexes
+#
+#  index_teacher_notifications_on_user_id  (user_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (user_id => users.id)
+#
+class TeacherNotification < ApplicationRecord
+  NOTIFICATION_TYPES = [
+    STUDENT_COMPLETED_DIAGNOSTIC = 'student-completed-diganostic',
+    STUDENT_COMPLETED_ALL_DIAGNOSTIC_RECOMMENDATIONS = 'student-completed-all-diagnostic-recommendations',
+    STUDENT_COMPLETED_ALL_ASSIGNED_ACTIVITIES = 'student-completed-all-assigned-activities'
+  ]
+
+  NOTIFICATION_TYPE_VALIDATION = {
```

Comment:
So I'm not opposed to this setup, but some Rails-y things to consider using instead. 

Consider using [store](https://api.rubyonrails.org/classes/ActiveRecord/Store.html). It's a way to store the json, but treat them as proper attributes. You could then use standard Rails validations, e.g. `validates :student_name, presence: true` would work. You can make validations conditional with `if:` clauses if needed. Also code such as `notification.student_name = "Thomas"` would work as expected.

Reply:
I was originally struggling to figure out how to use `store` in conjunction with the fact that different notification types had different shapes, but it turns out that the solution was just to use the STI pattern.  So I've refactored the models in a way that I think takes advantage of both suggestions.  Let me know if this looks like a good solution.